### PR TITLE
Convert CI jobs to use Gradle build

### PR DESCRIPTION
**Pull request by cljohnso**
The primary purpose of this commit is to change the use of Maven to Gradle for CI jobs.  This commit also:

* Removes `ci-javadoc.yaml` -- a change in the PlantUML plugin permits building Javadoc under Java 8+
* Introduces `propagateUserHome.gradle.kts` for use as a Gradle init script to "fix-up" a bad `user.home` value observed under container-based builds
* Adds suppression for Gradle POM metadata warnings
* Changes the use of `InvalidUserCodeException` to `InvalidUserDataException`

**Once this PR is vetted and merged, a subsequent PR will remove the Maven build components.**

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -44,16 +43,16 @@ jobs:
           distribution: ${{matrix.distribution}}
       - name: Install lsof
         run: yum -y install lsof procps
-      - name: Test with Maven
-        run: ./mvnw clean verify -V -B "-Dmaven.artifact.threads=64" "-Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn" -DskipToolchain "-Djava.version=${{matrix.java}}" "-Dmaven.javadoc.skip"
+      - name: Test with Gradle
+        run: ./gradlew -V clean check
```
Comment (chrisdennis) [inline .github/workflows/ci.yaml:47]: The clean shouldn't be needed... should be a no-op.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -76,20 +75,22 @@ jobs:
...
-        # Details: https://bugs.openjdk.java.net/browse/JDK-8193433
-        run: export MAVEN_OPTS=-Duser.home=$HOME && export JAVA_OPTS=-Duser.home=$HOME && ./mvnw clean verify -V -B "-Dmaven.artifact.threads=64" "-Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn" -DskipToolchain "-Djava.version=${{matrix.java}}" "-Dmaven.javadoc.skip"
+      - name: Test with Gradle
+        # When running in a container using an "anonymous" non-root user, 'user.home' is likely not set to a usable value
+        # and must be set manually. (See https://bugs.openjdk.java.net/browse/JDK-8193433)  Since this task uses an
+        # anonymous non-root user, a Gradle init script is used to set 'user.home' to "$HOME" and amend the testing tasks
+        # to pass that value through to the testing JVMs.
+        run: >
+          java -XshowSettings:all -version &&
+          ./gradlew -V --init-script .github/workflows/propagateUserHome.gradle.kts clean check
```
Comment (chrisdennis) [inline .github/workflows/ci.yaml:85]: `--show-version` instead of `-V`?

File: .github/workflows/propagateUserHome.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    /*
+     * Replace Java's broken user.home value in both this JVM and for each test JVM.
+     */
+    println("Setting user.home=\"${home}\"")
+    System.setProperty("user.home", home)
+
+    gradle.allprojects {
+        tasks.withType<Test> {
```
Comment (chrisdennis) [inline .github/workflows/propagateUserHome.gradle.kts:19]: You might be able to use:
```kotlin
tasks.withType<JavaForkOptions> {
}
```
which will catch both `Test` and `JavaExec` tasks.

File: .github/workflows/propagateUserHome.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            println("Setting ${this.path} user.home=\"${home}\"")
```
Comment (chrisdennis) [inline .github/workflows/propagateUserHome.gradle.kts:20]:
```
logger.info(...)
```

File: .github/workflows/propagateUserHome.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+if (userHome.isNullOrEmpty() || userHome.equals("?")) {
+    val home = System.getenv("HOME")
+    if (home.isNullOrEmpty()) {
+        throw IllegalStateException("environment variable \"HOME\" must be set when user.home='${userHome}'")
+    }
+
+    /*
+     * Replace Java's broken user.home value in both this JVM and for each test JVM.
+     */
+    println("Setting user.home=\"${home}\"")
```
Comment (chrisdennis) [inline .github/workflows/propagateUserHome.gradle.kts:15]: This should really go through a Gradle logger... I think at this layer you have to look it up through slf4j though.

File: .github/workflows/publish-pr-test-results.yaml [hunk 6]
```diff
@@ -35,4 +35,5 @@ jobs:
           commit: ${{ github.event.workflow_run.head_sha }}
           event_file: artifacts/Event File/event.json
           event_name: ${{ github.event.workflow_run.event }}
-          files: "**/target/surefire-reports/**/*.xml"
\ No newline at end of file
+          files: |
```
Comment (chrisdennis) [inline .github/workflows/publish-pr-test-results.yaml:38]: Why breaking over the line suddenly?

File: test-tools/build.gradle.kts [hunk 7]
```diff
@@ -136,6 +136,10 @@ tasks.sourcesJar {
 
 publishing {
     publications.withType<MavenPublication> {
+        suppressPomMetadataWarningsFor("apiElements")
```
Comment (chrisdennis) [inline test-tools/build.gradle.kts:139]: Why suppress suddenly?

**Review by chrisdennis — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (cljohnso) [reply .github/workflows/ci.yaml:47]: Residue of the Maven version -- removing.

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (cljohnso) [reply .github/workflows/ci.yaml:85]: K

File: .github/workflows/propagateUserHome.gradle.kts (hunk 3 above)
Comment (cljohnso) [reply .github/workflows/propagateUserHome.gradle.kts:19]: I'll try it and see what happens.

File: .github/workflows/propagateUserHome.gradle.kts (hunk 4 above)
Comment (cljohnso) [reply .github/workflows/propagateUserHome.gradle.kts:20]: I'll change to that once I get any kinks worked out -- right now, I just want to see it.

File: .github/workflows/propagateUserHome.gradle.kts (hunk 5 above)
Comment (cljohnso) [reply .github/workflows/propagateUserHome.gradle.kts:15]: I just found out that `println` is sent to `logger.quiet` ...

File: .github/workflows/publish-pr-test-results.yaml (hunk 6 above)
Comment (cljohnso) [reply .github/workflows/publish-pr-test-results.yaml:38]: I was considering multiple paths.  (At least now, it's consistent with the other, similar clauses in the yaml files.)

File: test-tools/build.gradle.kts (hunk 7 above)
Comment (cljohnso) [reply test-tools/build.gradle.kts:139]: No reason for the noise; I'm surprised it's not suppressed in the other places we use this construct -- we're aware of the gap and are satisfied with the result we get ... the extra warnings are useless.

**Review by chrisdennis — APPROVED** [review]

**Review by chrisdennis — APPROVED** [review]

**Pull request merged**